### ISSUE TITLE
Fix issue with user not saving in between changes

### ISSUE THREAD

**Issue by dEhiN** [dEhiN/Demord#24]
While working on issue #5, noted an edge case which basically involves when a user makes a change to startup data but doesn't save, and then makes another change. Creating this issue to work on that.

**Comment by dEhiN** [comment]
Copying the details from commit 63216ec4 (with some modifications) to explain the exact edge case discovered:

_The testing was done against the default startup data and comprised the following steps:_

_1) Edited startup item 2 to remove an argument, change the startup item name, and change the startup item description._

_2) Saved the modified startup item to disk. This was then copied to the JSON file "edge-case-one_startup_data.json"._

_3) Edited startup item 3 to add an argument, change the startup item name, and change the startup item description._

_4) Deleted startup item 1, so default startup item 2 became item 1, and default startup item 3 became item 2._

_5) Selected to edit startup item 2 from the udpated json_editor menu._

_6) Saved the startup item data to disk. This was then copied to the JSON file "edge-case-two_startup_data.json"._

_Two new JSON files were created by copying the saved data at steps 2 and then 6. This is because the JSON data saved at step 6 actually has 3 startup items, with items 1 and 3 being the exact same as the default startup data. However, since at step 5, the startup item selected was at that point startup item 2, in step 6, the default startup item 2 was replaced by this version of startup item 2._

_In other words, if you compare the two edge case startup JSON files, you'll see the edge case happens when a startup item is modified but not saved, the user returns to the previous menu, deletes a different startup item but doesn't save the full startup data, then selects to edit the same startup item (which now has a different startup item number) and saves that item to disk._

**Comment by dEhiN** [comment]
Thought of two possible ways to fix this:

1. Use some sort of variable to track when a change has been made to the startup data
2. Have every function that allows for changes to the startup data call the save function automatically before returning execution

For option 1, the variable would have to be some sort of flag that is used globally across the whole application. Could possibly create a 'helper' function in the `helpers` module that updates the flag. For example, when a user edits a startup item, the function `edit_startup_item` could then call this helper function to update the flag. If the user selects to save the modified startup item, this helper function is called again to update the flag. If the user instead selects to delete a startup item, this flag is checked and if the modified data wasn't saved, it's first saved. Alternatively, a message is displayed alerting the user and asking if they want to save the changes made so far.

Benefit of this option is that the user isn't forced to save every change they make, especially since there is currently no "undo" functionality. So, if a user makes changes to a startup item, but then decides to delete the item, they can choose to not save the changes made since they will be deleting the item anyway.

Downside to this option is that every function that deals with modifying the startup data as well as every save function will need to check with the helper function first. The helper function, in addition to checking for and updating the global flag, would probably need to be the one that alerts the user of unsaved changes and asks the user if they want to save, and then call one of the save functions.

For option 2, a function like `edit_startup_item` would either call a save function automatically before returning execution, or would call a 'helper' function. If the latter, this helper function would be similar to the one described in option 1 but there wouldn't be a need for a global flag. Instead, the helper function will ask the user if they want to save the changes. This would only be called, in this example, when the user selects to "Return to the previous menu". Of course, if the user had first selected to "Save startup item data to disk", then this helper function wouldn't be called.

Benefit of this option is that it will probably involve less code changes as there would be less places to refactor.

Downside of this option is that it could possibly be more bothersome to the user since every time they change a startup item, or delete a startup item, they will be asked to save. 

Another possibility is a similar approach to option 2, but with a slight twist. When a change is made and it's noted that the user is wanting to return to the previous menu without saving the changes, the user could be asked if they want to save or discard the changes. This way, if the user chooses to discard, the code can "undo" the changes by reverting to the original JSON data. The code could even just read in the JSON data, since every call to save writes the changes to the startup_data.json file.

For example:
1. User starts with default startup data of 3 startup items
2. User modifies item 2
3. User saves the modified item 2
4. User decides to delete item 3
5. User doesn't save the data but returns to the previous menu
6. User is presented with option to save or discard
7. User wanted to undo deletion which was why they returned to the previous menu
8. As a result, user chooses to discard changes
9. The startup_data.json file will still have the modifications made to item 2

**Comment by dEhiN** [comment]
Updating the list of options for a fix to include three more options:

1. Use some sort of variable to track when a change has been made to the startup data
2. Have every function that allows for changes to the startup data call the save function automatically before returning execution
3. Have a function that asks the user to save or discard any changes and call this function whenever the user wants to return to the previous menu without saving first
4. Automatically save every change, minor or major
5. Automatically save changes when returning to previous menu

Option 3 is discussed in the previous comment. Option 4 is unlikely to be implemented as this is forcing the user to not make any mistakes. Option 5 could be helpful as the user doesn't have to worry about remembering to save. However, a user may not want to save changes they made.

Most likely I will go with option 3, but will need to think some more on this.

**Comment by dEhiN** [comment]
Created a issue branch for this and read through everything to familiarize myself with what's going on. The case specified is an edge case, but it points to a deeper functionality issue:

There is no auto-save type functionality, so a user could make changes and if they forget to save those changes, they could lose all the changes.

The comment from March 26 lists 5 possible fixes to this issue. Will think on what may work best with the current code, since a lot of changes have been made since March 26. Will also reproduce the steps outlined for the edge case and see if there's still an issue. However, even if the edge case isn't present anymore, this saving issue still needs to be addressed.

**Comment by dEhiN** [comment]
For now, decided to just have a note reminder to the user to save any changes they made before returning to the previous menu. This will need to be done in 3 places:

1) The menu "Edit the starutp file"
2) The menu "Edit startup item #"
3) The menu "Add new startup items"
